src/cmd/ksh93/sh/jobs.c

```diff
@@ -598,40 +597,17 @@ void job_init(Shell_t *shp, int lflag) {
 #endif /* NTTYDISC */
 
     job.mypgid = getpgrp();
-    // Some systems have job control, but not initialized.
-    if (job.mypgid <= 0) {
-        // Get a controlling terminal and set process group.
-        // This should have already been done by rlogin.
-        int fd;
-        char *ttynam;
-        int err = errno;
-#ifndef SIGTSTP
-        setpgid(0, shp->gd->pid);
-#endif  // SIGTSTP
-        if (job.mypgid < 0 || !(ttynam = ttyname(JOBTTY))) return;
-        while (close(JOBTTY) < 0 && errno == EINTR) errno = err;
-        if ((fd = sh_open(ttynam, O_RDWR)) < 0) return;
-        if (fd != JOBTTY) sh_iorenumber(shp, fd, JOBTTY);
-        job.mypgid = shp->gd->pid;
-#ifdef SIGTSTP
-        tcsetpgrp(JOBTTY, shp->gd->pid);
-        setpgid(0, shp->gd->pid);
-#endif  // SIGTSTP
-    }
 #ifdef SIGTSTP
-    possible = setpgid(0, job.mypgid) >= 0;
```

Comment:
This call to `setpgid()` succeeds in last stable version of ksh. I am still checking what's causing it to fail now.

Reply:
`setpgid()` is actually failing on last stable version of ksh. This bug was introduced by reformatting changes in 00d0a53b8.

This line:
```
    if(possible = (setpgid(0,job.mypgid)>=0) || errno==EPERM)     
```
was changed to:

```
   possible = setpgid(0, job.mypgid) >= 0;
    if (possible || errno == EPERM) {
```
So `possible` is not set to 1 if `setpgid()` fails due to `EPERM`. It should have been changed to:

```
   possible = setpgid(0, job.mypgid) >= 0  || errno == EPERM;
    if (possible) {
```

Reply:
Good catch and this is a good example for why the existing style that does assignment inside `if()` conditions is a really bad idea. Also, the `>= 0` test is itself confusing since `setpgid()` by definition returns zero on success, -1 on failure, and should never return a positive value. It would be clearer to write it as

```c
possible = setpgid(0, job.mypgid) != -1 || errno == EPERM;
```